modules/percolator/src/main/java/org/elasticsearch/percolator/QueryAnalyzer.java

```diff
@@ -564,35 +579,37 @@ private static Result handleDisjunction(List<Query> disjunctions, int requiredSh
             int resultMsm = subResult.minimumShouldMatch;
             for (QueryExtraction extraction : subResult.extractions) {
                 if (terms.add(extraction) == false) {
-                    resultMsm = Math.max(0, resultMsm - 1);
+                    verified = false;
+                    hasDuplicateTerms = true;
                 }
             }
-            clauses.add(new DisjunctionClause(resultMsm, subResult.extractions.stream()
-                .filter(extraction -> extraction.range != null)
-                .map(extraction -> extraction.range.fieldName)
-                .collect(toSet())));
+            if (hasRangeExtractions == false) {
+                hasRangeExtractions = subResult.extractions.stream().anyMatch(qe -> qe.range != null);
+            }
+            clauses.add(resultMsm);
         }
         boolean matchAllDocs = numMatchAllClauses > 0 && numMatchAllClauses >= requiredShouldClauses;
 
         int msm = 0;
-        if (version.onOrAfter(Version.V_6_1_0)) {
-            Set<String> seenRangeFields = new HashSet<>();
+        if (version.onOrAfter(Version.V_6_1_0) &&
+            // Having ranges would mean we need to juggle with the msm and that complicates this logic a lot,
+            // so for now lets not do it.
+            hasRangeExtractions == false) {
             // Figure out what the combined msm is for this disjunction:
             // (sum the lowest required clauses, otherwise we're too strict and queries may not match)
             clauses = clauses.stream()
-                .filter(o -> o.msm > 0)
-                .sorted(Comparator.comparingInt(o -> o.msm))
+                .filter(val -> val > 0)
+                .sorted()
                 .collect(Collectors.toList());
-            int limit = Math.min(clauses.size(), Math.max(1, requiredShouldClauses));
-            for (int i = 0; i < limit; i++) {
-                if (clauses.get(i).rangeFieldNames.isEmpty() == false) {
-                    for (String rangeField: clauses.get(i).rangeFieldNames) {
-                        if (seenRangeFields.add(rangeField)) {
-                            msm += 1;
-                        }
-                    }
-                } else {
-                    msm += clauses.get(i).msm;
+
+            // When there are duplicated query extractions, percolator can no longer reliably determine msm across this disjunction
+            if (hasDuplicateTerms) {
+                // pick lowest msm:
+                msm = clauses.get(0);
```

Comment:
is there always at least one element in this list? (I haven't checked whether we assert it somewhere else)

Reply:
Yes, if `hasDuplicateTerms` has been set to `true` then there should be at least two clauses.